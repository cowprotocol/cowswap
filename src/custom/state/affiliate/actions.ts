import { createAction } from '@reduxjs/toolkit'

export const updateReferralAddress = createAction<{
  value: string
  isValid: boolean
} | null>('affiliate/updateReferralAddress')

export const setReferralAddressActive = createAction<boolean>('affiliate/setReferralAddressActive')

export const updateAddress = createAction<string>('affiliate/updateAddress')

export const dismissNotification = createAction<string>('affiliate/dismissNotification')
