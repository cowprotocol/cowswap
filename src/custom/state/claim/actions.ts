import { createAction } from '@reduxjs/toolkit'
import { SupportedChainId } from 'constants/chains'
import { ClaimInfo } from 'state/claim/reducer'

export enum ClaimStatus {
  DEFAULT = 'DEFAULT',
  ATTEMPTING = 'ATTEMPTING',
  SUBMITTED = 'SUBMITTED',
  CONFIRMED = 'CONFIRMED',
  FAILED = 'FAILED',
}

export type ClaimActions = {
  // account
  setInputAddress: (payload: string) => void
  setActiveClaimAccount: (payload: string) => void
  setActiveClaimAccountENS: (payload: string) => void

  // search
  setIsSearchUsed: (payload: boolean) => void

  // claiming
  setClaimStatus: (payload: ClaimStatus) => void
  setClaimedAmount: (payload: string) => void
  setEstimatedGas: (payload: string) => void

  // investing
  setIsInvestFlowActive: (payload: boolean) => void
  setInvestFlowStep: (payload: number) => void
  initInvestFlowData: () => void
  updateInvestAmount: (payload: { index: number; amount: string }) => void
  updateInvestError: (payload: { index: number; error: string | undefined }) => void
  setIsTouched: (payload: { index: number; isTouched: boolean }) => void

  // claim row selection
  setSelected: (payload: number[]) => void
  setSelectedAll: (payload: boolean) => void
}

// accounts
export const setInputAddress = createAction<string>('claim/setInputAddress')
export const setActiveClaimAccount = createAction<string>('claim/setActiveClaimAccount')
export const setActiveClaimAccountENS = createAction<string>('claim/setActiveClaimAccountENS')

// search
export const setIsSearchUsed = createAction<boolean>('claim/setIsSearchUsed')

// claiming
export const setClaimedAmount = createAction<string>('claim/setClaimedAmount')
export const setClaimStatus = createAction<ClaimStatus>('claim/setClaimStatus')
export const setEstimatedGas = createAction<string>('claim/setEstimatedGas')

// investing
export const setIsInvestFlowActive = createAction<boolean>('claim/setIsInvestFlowActive')
export const setInvestFlowStep = createAction<number>('claim/setInvestFlowStep')
export const initInvestFlowData = createAction('claim/initInvestFlowData')
export const updateInvestAmount = createAction<{
  index: number
  amount: string
}>('claim/updateInvestAmount')
export const updateInvestError = createAction<{
  index: number
  error: string | undefined
}>('claim/updateInvestError')
export const setIsTouched = createAction<{
  index: number
  isTouched: boolean
}>('claim/setIsTouched')
// claim row selection
export const setSelected = createAction<number[]>('claim/setSelected')
export const setSelectedAll = createAction<boolean>('claim/setSelectedAll')
// Claim UI reset sugar
export const resetClaimUi = createAction('claims/resetClaimUi')
// Claims on other chains
export const setClaimsCount = createAction<{
  chain: SupportedChainId
  claimInfo: Partial<ClaimInfo>
  account: string
}>('claims/setClaimsCount')
