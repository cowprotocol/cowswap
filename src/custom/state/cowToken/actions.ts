import { createAction } from '@reduxjs/toolkit'

export enum SwapVCowStatus {
  INITIAL = 'INITIAL',
  ATTEMPTING = 'ATTEMPTING',
  SUBMITTED = 'SUBMITTED',
}

export type CowTokenActions = {
  setSwapVCowStatus: (payload: SwapVCowStatus) => void
}

export const setSwapVCowStatus = createAction<SwapVCowStatus>('cowToken/setSwapVCowStatus')
