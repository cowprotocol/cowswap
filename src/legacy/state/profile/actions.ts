import { createAction } from '@reduxjs/toolkit'

export const closeAnnouncement = createAction<{ contentHash: string }>('profile/closeAnnouncement')
