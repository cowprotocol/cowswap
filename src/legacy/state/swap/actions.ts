import { createAction } from '@reduxjs/toolkit'

export enum Field {
  INPUT = 'INPUT',
  OUTPUT = 'OUTPUT',
}

export interface ReplaceOnlyTradeRawStatePayload {
  readonly chainId: number | null
  readonly inputCurrencyId?: string
  readonly outputCurrencyId?: string
  readonly recipient: string | null
}

export interface ReplaceSwapStatePayload extends ReplaceOnlyTradeRawStatePayload {
  readonly typedValue: string
  readonly independentField: Field
}

export const selectCurrency = createAction<{ field: Field; currencyId: string }>('swap/selectCurrency')
export const switchCurrencies = createAction<void>('swap/switchCurrencies')
export const typeInput = createAction<{ field: Field; typedValue: string }>('swap/typeInput')

/**
 * Replaces only a subset of the state corresponding to TradeRawState.
 */
export const replaceOnlyTradeRawState = createAction<ReplaceOnlyTradeRawStatePayload>('swap/replaceOnlyTradeRawState')
export const replaceSwapState = createAction<ReplaceSwapStatePayload>('swap/replaceSwapState')
export const setRecipient = createAction<{ recipient: string | null }>('swap/setRecipient')
