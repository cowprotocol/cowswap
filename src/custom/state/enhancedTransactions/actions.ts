import { SafeMultisigTransactionResponse } from '@gnosis.pm/safe-service-client'
import { createAction } from '@reduxjs/toolkit'
import { SerializableTransactionReceipt } from '@src/state/transactions/actions'
import { EnhancedTransactionDetails } from './reducer'

type WithChainId = { chainId: number }
type WithData = { data?: any }

export type AddTransactionParams = WithChainId &
  WithData &
  Pick<
    EnhancedTransactionDetails,
    | 'hash'
    | 'hashType'
    | 'from'
    | 'approval'
    | 'presign'
    | 'claim'
    | 'summary'
    | 'safeTransaction'
    | 'swapVCow'
    | 'swapLockedGNOvCow'
  >

export const addTransaction = createAction<AddTransactionParams>('enhancedTransactions/addTransaction')

export const clearAllTransactions = createAction<WithChainId>('enhancedTransactions/clearAllTransactions')

export const finalizeTransaction = createAction<{
  chainId: number
  hash: string
  receipt: SerializableTransactionReceipt
  safeTransaction?: SafeMultisigTransactionResponse
}>('enhancedTransactions/finalizeTransaction')

export const checkedTransaction = createAction<{
  chainId: number
  hash: string
  blockNumber: number
}>('enhancedTransactions/checkedTransaction')

export type ReplacementType = 'speedup' | 'cancel'

export const replaceTransaction = createAction<{
  chainId: number
  oldHash: string
  newHash: string
  type: ReplacementType
}>('enhancedTransactions/replaceTransaction')

export const updateSafeTransaction = createAction<{
  chainId: number
  safeTransaction: SafeMultisigTransactionResponse
  blockNumber: number
}>('enhancedTransactions/updateSafeTransaction')
