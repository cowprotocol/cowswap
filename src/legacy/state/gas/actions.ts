import { createAction } from '@reduxjs/toolkit'
import { WithChainId } from 'legacy/state/lists/actions'
import { GasFeeEndpointResponse } from 'api/gasPrices'

export type UpdateGasPrices = GasFeeEndpointResponse & WithChainId

export const updateGasPrices = createAction<UpdateGasPrices>('gas/updateGasPrices')
