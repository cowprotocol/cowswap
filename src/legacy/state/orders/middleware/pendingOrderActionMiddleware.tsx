import { getOrderByIdFromState, OrderTxTypes, PopupPayload, setPopupData } from '../helpers'
import { orderAnalytics } from '../../../components/analytics'
import { AddPendingOrderParams } from '../actions'
import { MiddlewareAPI } from '@reduxjs/toolkit'
import { AppState } from '../../index'
import { Dispatch } from 'redux'
import { addPopup } from '../../application/reducer'

export function pendingOrderActionMiddleware(
  store: MiddlewareAPI<Dispatch, AppState>,
  payload: AddPendingOrderParams,
  result: any
) {
  const { id, chainId } = payload

  // use current state to lookup orders' data
  const orders = store.getState().orders[chainId]
  const orderObject = getOrderByIdFromState(orders, id)

  if (!orderObject) {
    return result
  }
  // look up Order.summary for Popup
  const { summary, class: orderClass } = orderObject.order

  let popup: PopupPayload

  const hash = orderObject.order.orderCreationHash
  if (hash) {
    // EthFlow Order
    popup = setPopupData(OrderTxTypes.TXN, {
      summary,
      status: 'submitted',
      id: hash,
      hash,
    })
    orderAnalytics('Posted', orderClass, 'EthFlow')
  } else if (!payload.order.isHidden) {
    // Pending Order Popup, if it's not hidden
    popup = setPopupData(OrderTxTypes.METATXN, { summary, status: 'submitted', id })
    orderAnalytics('Posted', orderClass, 'Offchain')
  }

  if (popup) {
    store.dispatch(
      addPopup({
        id,
        popup,
      })
    )
  }
}
