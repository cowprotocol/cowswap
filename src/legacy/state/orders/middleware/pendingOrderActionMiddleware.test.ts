import { pendingOrderActionMiddleware } from './pendingOrderActionMiddleware'
import { MiddlewareAPI } from '@reduxjs/toolkit'
import { AnyAction, Dispatch } from 'redux'
import { AppState } from '../../index'
import { AddPendingOrderParams } from '../actions'
import { OrderClass } from '@cowprotocol/cow-sdk'
import { anything, capture, instance, mock, resetCalls, verify, when } from 'ts-mockito'

const MOCK_ETHFLOW_ORDER = {
  '0x001': {
    id: '0x001',
    order: {
      summary: 'summary',
      orderClass: OrderClass.MARKET,
      orderCreationHash: '0xhash',
    },
  },
}

const MOCK_REGULAR_ORDER = {
  '0x002': {
    id: '0x002',
    order: {
      summary: 'summary',
      orderClass: OrderClass.LIMIT,
    },
  },
}

const MOCK_ORDERS_STORE = {
  1: {
    pending: { ...MOCK_ETHFLOW_ORDER, ...MOCK_REGULAR_ORDER },
  },
}

const storeMock = mock<MiddlewareAPI<Dispatch, AppState>>()
const payloadMock = mock<AddPendingOrderParams>()

describe('pendingOrderActionMiddleware', () => {
  when(storeMock.getState()).thenReturn({ orders: MOCK_ORDERS_STORE } as any)
  when(payloadMock.chainId).thenReturn(1)

  beforeEach(() => {
    resetCalls(storeMock)
    resetCalls(payloadMock)
  })

  it('should not trigger pop up for inexistent order', () => {
    when(payloadMock.id).thenReturn('0x000')

    pendingOrderActionMiddleware(instance(storeMock), instance(payloadMock))

    verify(storeMock.dispatch(anything())).never()
  })
  it('should trigger pop up for ethflow order', () => {
    when(payloadMock.id).thenReturn('0x001')

    pendingOrderActionMiddleware(instance(storeMock), instance(payloadMock))

    const [addPopupAction] = capture(storeMock.dispatch<AnyAction>).first()

    expect(addPopupAction).toMatchSnapshot()
  })
  it('should trigger pop up for regular order', () => {
    when(payloadMock.id).thenReturn('0x002')
    // @ts-ignore
    when(payloadMock.order).thenReturn({ isHidden: false })

    pendingOrderActionMiddleware(instance(storeMock), instance(payloadMock))

    const [addPopupAction] = capture(storeMock.dispatch<AnyAction>).first()

    expect(addPopupAction).toMatchSnapshot()
  })
  it('should not trigger pop up for hidden order', () => {
    when(payloadMock.id).thenReturn('0x003')
    // @ts-ignore
    when(payloadMock.order).thenReturn({ isHidden: true })

    pendingOrderActionMiddleware(instance(storeMock), instance(payloadMock))

    verify(storeMock.dispatch(anything())).never()
  })
})
