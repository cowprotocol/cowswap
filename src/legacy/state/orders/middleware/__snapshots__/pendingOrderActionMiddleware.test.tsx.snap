// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pendingOrderActionMiddleware When some conditions, then should dispatch addPopup event 1`] = `
Object {
  "payload": Object {
    "id": "0x001",
    "popup": Object {
      "content": Object {
        "txn": Object {
          "hash": "0xhash",
          "success": true,
          "summary": "summary submitted",
        },
      },
      "key": "0xhash_submitted",
    },
  },
  "type": "application/addPopup",
}
`;
